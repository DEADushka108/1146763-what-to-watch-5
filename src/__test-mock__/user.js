const authorizationStatus = `200`;

const userInfo = {
  id: 1,
  name: `batman`,
  email: `batman@batman.net`,
  avatar: `https://assets.htmlacademy.ru/intensives/javascript-3/avatar/4.jpg`,
};

export {authorizationStatus, userInfo};
