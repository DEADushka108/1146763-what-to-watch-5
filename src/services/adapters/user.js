const createUserInfo = (user) => ({
  id: user.id,
  name: user.name,
  email: user.email,
  avatar: user[`avatar_url`],
});

export {createUserInfo};
