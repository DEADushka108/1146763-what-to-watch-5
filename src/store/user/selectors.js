import {createSelector} from 'reselect';
import {AuthorizationStatus} from '../../utils/const.js';
import NameSpace from '../name-space.js';

const getAuthorizationStatus = (state) => {
  return state[NameSpace.USER].authorizationStatus;
};

const getFavoriteList = (state) => {
  return state[NameSpace.USER].favoriteList;
};

const getUserInfo = (state) => {
  return state[NameSpace.USER].userInfo;
};

const getLoginStatus = (state) => {
  return state[NameSpace.USER].loginStatus;
};

const getAuthorization = createSelector(
    getAuthorizationStatus,
    (status) => {
      return status === AuthorizationStatus.AUTH;
    }
);

export {getUserInfo, getAuthorizationStatus, getFavoriteList, getAuthorization, getLoginStatus};
