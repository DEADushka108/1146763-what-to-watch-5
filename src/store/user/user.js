import {extend} from '../../utils/utils.js';
import {URL, AuthorizationStatus, AppRoute} from '../../utils/const.js';
import {createMoviesList} from '../../services/adapters/movies.js';
import {createUserInfo} from '../../services/adapters/user.js';
import {redirectToRoute} from '../redirect/redirect-action.js';

const initialState = {
  authorizationStatus: AuthorizationStatus.NO_AUTH,
  favoriteList: [],
  userInfo: {
    id: null,
    name: null,
    email: null,
    avatar: null,
  },
};

const ActionType = {
  REQUIRE_AUTHORIZATION: `REQUIRE_AUTHORIZATION`,
  LOAD_FAVORITE_LIST: `LOAD_FAVORITE_LIST`,
  GET_USER_INFO: `GET_USER_INFO`,
};

const ActionCreator = {
  requireAuthorization: (status) => ({
    type: ActionType.REQUIRE_AUTHORIZATION,
    payload: status,
  }),
  loadFavoriteList: (list) => ({
    type: ActionType.LOAD_FAVORITE_LIST,
    payload: list,
  }),
  getUserInfo: (userInfo) => ({
    type: ActionType.GET_USER_INFO,
    payload: userInfo,
  })
};

const Operation = {
  checkAuthorization: () => (dispatch, _getState, api) => {
    return api.get(URL.LOGIN)
      .then((response) => {
        dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
        dispatch(ActionCreator.getUserInfo(createUserInfo(response.data)));
      })
      .catch(() => {
        dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.NO_AUTH));
      });
  },
  login: (userData) => (dispatch, _getState, api) => {
    return api.post(URL.LOGIN, {
      email: userData.login,
      password: userData.password,
    })
      .then((response) => {
        dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
        dispatch(ActionCreator.getUserInfo(createUserInfo(response.data)));
        dispatch(redirectToRoute(AppRoute.ROOT));
      });
  },
  loadFavoriteList: () => (dispatch, _getState, api) => {
    return api.get(URL.FAVORITE)
      .then((response) => {
        dispatch(ActionCreator.loadFavoriteList(createMoviesList(response.data)));
      });
  },
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.REQUIRE_AUTHORIZATION:
      return extend(state, {
        authorizationStatus: action.payload,
      });
    case ActionType.LOAD_FAVORITE_LIST:
      return extend(state, {
        favoriteList: action.payload,
      });
    case ActionType.GET_USER_INFO:
      return extend(state, {
        userInfo: action.payload,
      });
  }

  return state;
};

export {reducer, ActionCreator, ActionType, Operation};
