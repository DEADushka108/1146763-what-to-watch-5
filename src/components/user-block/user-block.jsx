import React from 'react';
import PropTypes from 'prop-types';
import {AppRoute} from '../../utils/const';
import {Link} from 'react-router-dom';
import {connect} from 'react-redux';
import {getAuthorization, getUserInfo} from '../../store/user/selectors';
import {userDetails} from '../../types/types';

const UserBlock = (props) => {
  const {isAuthorized, userInfo} = props;
  const {avatar, name} = userInfo;

  return <div className="user-block">
    {isAuthorized ? <Link to={AppRoute.FAVORITE}><div className="user-block__avatar">
      <img src={avatar} alt={name} width="63" height="63"/>
    </div></Link> : <Link to={AppRoute.LOGIN} className="user-block__link">Sign in</Link>}
  </div>;
};

UserBlock.propTypes = {
  isAuthorized: PropTypes.bool.isRequired,
  userInfo: userDetails,
};

const mapStateToProps = (state) => ({
  isAuthorized: getAuthorization(state),
  userInfo: getUserInfo(state),
});

export {UserBlock};
export default connect(mapStateToProps)(UserBlock);
