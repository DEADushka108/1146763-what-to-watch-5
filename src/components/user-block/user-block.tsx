import React from 'react';
import {AppRoute} from '../../utils/const';
import {Link} from 'react-router-dom';
import {connect} from 'react-redux';
import {getAuthorization, getUserInfo} from '../../store/user/selectors';
import {User} from '../../types/user';

interface Props {
  isAuthorized: boolean;
  userInfo: User;
}

const UserBlock = (props: Props) => {
  const {isAuthorized, userInfo} = props;
  const {avatar, name} = userInfo;

  return <div className="user-block">
    {isAuthorized ?
      <Link to={AppRoute.FAVORITE}>
        <div className="user-block__avatar">
          <img src={avatar} alt={name} width="63" height="63"/>
        </div>
      </Link> :
      <Link to={AppRoute.LOGIN} className="user-block__link">
        Sign in
      </Link>}
  </div>;
};

const mapStateToProps = (state) => ({
  isAuthorized: getAuthorization(state),
  userInfo: getUserInfo(state),
});

export {UserBlock};
export default connect(mapStateToProps)(UserBlock);
