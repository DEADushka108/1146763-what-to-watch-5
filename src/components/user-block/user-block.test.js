import React from 'react';
import renderer from 'react-test-renderer';
import {MemoryRouter} from 'react-router-dom';
import {authorizationStatus, userInfo} from '../../__test-mock__/user.js';
import {UserBlock} from './user-block.jsx';

it(`Should render UserBlock component correctly`, () => {
  const tree = renderer
    .create(
        <MemoryRouter>
          <UserBlock
            authorizationStatus={authorizationStatus}
            isAuthorized={true}
            userInfo={userInfo}
          />
        </MemoryRouter>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
