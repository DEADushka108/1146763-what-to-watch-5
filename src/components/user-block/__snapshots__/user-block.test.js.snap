// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render UserBlock component correctly 1`] = `
<div
  className="user-block"
>
  <a
    href="/mylist"
    onClick={[Function]}
  >
    <div
      className="user-block__avatar"
    >
      <img
        alt="batman"
        height="63"
        src="https://assets.htmlacademy.ru/intensives/javascript-3/avatar/4.jpg"
        width="63"
      />
    </div>
  </a>
</div>
`;
